src/server/middleware/transaction/transaction.go

```diff
@@ -0,0 +1,84 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transaction
+
+import (
+	"context"
+	"errors"
+	"net/http"
+
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/internal"
+	"github.com/goharbor/harbor/src/internal/orm"
+	"github.com/goharbor/harbor/src/server/middleware"
+)
+
+// Config defines the config for Transaction middleware.
+type Config struct {
+	// Acceptor defines a function to commit transaction
+	Acceptor func(statusCode int) bool
+}
+
+var (
+	// DefaultConfig default transaction config
+	DefaultConfig = Config{
+		// acceptor which will commit the transaction when status code between 200 to 399 as success
+		Acceptor: func(statusCode int) bool {
+			// statusCode between 200 to 399 as success
+			return statusCode >= http.StatusOK && statusCode < http.StatusBadRequest
+		},
+	}
+
+	errNonSuccess = errors.New("non success status code")
+)
+
+// Middleware middleware which add transaction for the http request with default config
+func Middleware(skippers ...middleware.Skipper) func(http.Handler) http.Handler {
+	return MiddlewareWithConfig(DefaultConfig, skippers...)
+}
+
+// MiddlewareWithConfig middleware which add transaction for the http request with config
+func MiddlewareWithConfig(config Config, skippers ...middleware.Skipper) func(http.Handler) http.Handler {
+	if config.Acceptor == nil {
+		config.Acceptor = DefaultConfig.Acceptor
+	}
+
+	return middleware.New(func(w http.ResponseWriter, r *http.Request, next http.Handler) {
+		res := internal.NewResponseBuffer(w)
+		defer res.Flush()
+
+		h := func(ctx context.Context) error {
+			next.ServeHTTP(res, r)
+
+			if !config.Acceptor(res.StatusCode()) {
```

Comment:
There is res.Success()

Reply:
Acceptor in the Config of the middleware need the status code, so that can user of the middleware can define their own logic to  decide which status code is successful.